### PR TITLE
[ IMPROVEMENT ] Support Caching of Video/Images

### PR DESCRIPTION
**Pull request by pratyaksh123**
## Proposed changes

Currently, there is no way to cache images/Videos. ( You can save it, but each time you open `AttachmentView.js` it will make a network request that makes the UX bad. This PR aims to solve this issue by storing these media assets in the user's Document Directory. This improves video load times to load almost instantaneously
## Issue(s)	
https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/2389

## How to test or reproduce

Play a video. ( After it is downloaded )

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

We need to make a decision as to how the caching behavior should be when a user opens a video -
- [ ] Auto-download the video if it is not downloaded yet. ( maintain a global setting that users can toggle that decides if to auto download or not , similar to WhatsApp. )
- [ ] Use the existing `HandleSave` method to allow users to save the video explicitly.

This PR will also make loading of Video Thumbnails pretty instantaneous :slightly_smiling_face: https://github.com/RocketChat/Rocket.Chat.ReactNative/pull/2772

### CODE REVIEW

**Comment by pratyaksh123** [comment]
@diegolmello what should be the caching behavior be like  ? ( check in the further comments above ) 
Please review when you get free.

File: app/views/AttachmentView.js [hunk 1]
```diff
@@ -146,22 +147,57 @@ class AttachmentView extends React.Component {
...
+			const files = await FileSystem.readDirectoryAsync(FileSystem.documentDirectory);
+			files.forEach((t) => {
+				if (t === fileName) {
+					this.setState({ path: `${ FileSystem.documentDirectory + fileName }` });
+				}
+			});
+		} catch (e) {
+			EventEmitter.emit(LISTENER, { message: I18n.t(image_url ? 'error-save-image' : 'error-save-video') });
+		}
+	}
```
Comment (gerzonc) [inline app/views/AttachmentView.js:180]: This function is pretty much like `handleSave`, can you refactor it so we can extract what we need for both?

File: app/views/AttachmentView.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ import RNFetchBlob from 'rn-fetch-blob';
 import { Video } from 'expo-av';
 import SHA256 from 'js-sha256';
 import { withSafeAreaInsets } from 'react-native-safe-area-context';
-
+import * as FileSystem from 'expo-file-system';
```
Comment (gerzonc) [inline app/views/AttachmentView.js:11]: Why not `rn-fetch-blob`?

**Review by gerzonc — CHANGES_REQUESTED** [review]

File: app/views/AttachmentView.js (hunk 2 above)
Comment (pratyaksh123) [reply app/views/AttachmentView.js:11]: `expo-FileSystem` had a more intuitive API and we were already using that so thought using this might be easier.

**Comment by pratyaksh123** [comment]
Also, can you check the further comments I made in the PR above, we need to decide when to cache media, on handleSave, or on opening it for the first time?

**Comment by raza2022** [comment]
@pratyaksh123 @gerzonc any update on this one as I just wanna it but some customize way but might finish it to here, instead of starting from scratch

**Comment by pratyaksh123** [comment]
> @pratyaksh123 @gerzonc any update on this one as I just wanna it but some customized way so thinking to finish it here, instead of starting from scratch

You can pick it up , I'm no longer working on rocket.chat issues

**Pull request closed without merging**